### PR TITLE
[SPARK-34568][SQL] When SparkContext's conf not enable hive, we should respect `enableHiveSupport()` when build SparkSession too

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
When SparkContext is initialed, if we want to start SparkSession, when we call 
`SparkSession.builder.enableHiveSupport().getOrCreate()`, the SparkSession we created won't have hive support since 
we have't reset existed SC's conf's `spark.sql.catalogImplementation`.
In this PR we use sharedState.conf to decide whether we should enable Hive Support.

### Why are the changes needed?
We should respect `enableHiveSupport`

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
Gentle ping @HyukjinKwon  Since I am not familiar with python and r, hope for some help about python and r.

**Comment by yaooqinn** [comment]
> When SparkContext is initialed, if we want to start SparkSession, when we call `SparkSession.builder.enableHiveSupport().getOrCreate()`, the SparkSession we created won't have hive support

Is it true? In this case `enableHiveSupport` will add `spark.sql.catalogImplementation` to the initial configs for `SparkSession` and `SharedState`. And then, the config will be populated in the cloned SparkConf of `SharedState`, which is used to init the externalCatalog.

**Comment by AngersZhuuuu** [comment]
> > When SparkContext is initialed, if we want to start SparkSession, when we call `SparkSession.builder.enableHiveSupport().getOrCreate()`, the SparkSession we created won't have hive support
> 
> Is it true? In this case `enableHiveSupport` will add `spark.sql.catalogImplementation` to the initial configs for `SparkSession` and `SharedState`. And then, the config will be populated in the cloned SparkConf of `SharedState`, which is used to init the externalCatalog.

But sessionStated is initialed according to SparkContext's conf
```
  lazy val sessionState: SessionState = {
    parentSessionState
      .map(_.clone(this))
      .getOrElse {
        val state = SparkSession.instantiateSessionState(
          SparkSession.sessionStateClassName(sparkContext.conf),
          self,
          initialSessionOptions)
        state
      }
  }
```

**Comment by yaooqinn** [comment]
```sql
           SparkSession.sessionStateClassName(sharedState.conf),
```

seems correct to me

**Comment by AngersZhuuuu** [comment]
> ```sql
>            SparkSession.sessionStateClassName(sharedState.conf),
> ```
> 
> seems correct to me

LGTM, I will try this way.

**Comment by AngersZhuuuu** [comment]
> ```sql
>            SparkSession.sessionStateClassName(sharedState.conf),
> ```
> 
> seems correct to me

Hmmmm 
```
private[spark] object HiveUtils extends Logging {

  def withHiveExternalCatalog(sc: SparkContext): SparkContext = {
    sc.conf.set(CATALOG_IMPLEMENTATION.key, "hive")
    sc
  }
```

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -946,7 +946,11 @@ object SparkSession extends Logging {
           SparkContext.getOrCreate(sparkConf)
           // Do not update `SparkConf` for existing `SparkContext`, as it's shared by all sessions.
         }
-
+        // We should reset `spark.sql.catalogImplementation` according to current requirement.
+        if (sparkContext.conf.get(CATALOG_IMPLEMENTATION) == "in-memory" &&
+          sparkConf.get(CATALOG_IMPLEMENTATION) == "hive") {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:951]: What if `sparkContext.conf.get(CATALOG_IMPLEMENTATION) == "hive" && sparkConf.get(CATALOG_IMPLEMENTATION) == "in-memory"`?

Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:951]: > What if `sparkContext.conf.get(CATALOG_IMPLEMENTATION) == "hive" && sparkConf.get(CATALOG_IMPLEMENTATION) == "in-memory"`?

I think if `sparkContext.conf.get(CATALOG_IMPLEMENTATION) == "hive"`, I think user have set this before start SparkSession.

And we have a method `HiveUtils. withHiveExternalCatalog` we should respect this too ==

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:951]: > I think if sparkContext.conf.get(CATALOG_IMPLEMENTATION) == "hive", I think user have set this before start SparkSession.

`CATALOG_IMPLEMENTATION` is a static configuration, so we can change the value that a user sets before starting a SparkContext?

Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:951]: > > I think if sparkContext.conf.get(CATALOG_IMPLEMENTATION) == "hive", I think user have set this before start SparkSession.
> 
> `CATALOG_IMPLEMENTATION` is a static configuration, so we can change the value that a user sets before starting a SparkContext?

But if we start SparkContext and then call `HiveUtils. withHiveExternalCatalog()` then use this SparkContext to start SparkSession?

Seems our test framework is build like this. and you can see my test result before the last finished one.

**Comment by yaooqinn** [comment]
> > ```sql
> >            SparkSession.sessionStateClassName(sharedState.conf),
> > ```
> > 
> > 
> > seems correct to me
> 
> Hmmmm
> 
> ```
> private[spark] object HiveUtils extends Logging {
> 
>   def withHiveExternalCatalog(sc: SparkContext): SparkContext = {
>     sc.conf.set(CATALOG_IMPLEMENTATION.key, "hive")
>     sc
>   }
> ```

Can't we use `SparkSession.builder().enableHiveSupport().sparkContext(sc).getOrCreate()` in `HiveContext`？

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
Gentle ping @maropu @yaooqinn Current way should be more suitable.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala [hunk 2]
```diff
@@ -34,7 +34,7 @@ class HiveContext private[hive](_sparkSession: SparkSession)
   self =>
 
   def this(sc: SparkContext) = {
-    this(SparkSession.builder().sparkContext(HiveUtils.withHiveExternalCatalog(sc)).getOrCreate())
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:37]: `withHiveExternalCatalog` is used only for tests after this merged, so could you move it into a suitable place?

Comment (AngersZhuuuu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:37]: > `withHiveExternalCatalog` is used only for tests after this merged, so could you move it into a suitable place?

Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala [hunk 3]
```diff
@@ -88,4 +89,17 @@ class HiveSharedStateSuite extends SparkFunSuite {
     assert(!ss.conf.get(WAREHOUSE_PATH).contains(invalidPath),
       "session level conf should be passed to catalog")
   }
+
+  test("SPARK-34568: We should respect enableHiveSupport when initialize SparkSession") {
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala:93]: It seems this test can pass in the current master. Have you checked that?

Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala:93]: > It seems this test can pass in the current master. Have you checked that?

Hmmm, make a mistake, we should check SessionState.., Updated

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 4]
```diff
@@ -685,3 +685,10 @@ private[hive] object HiveTestJars {
     targetFile
   }
 }
+
+private[hive] object TestHiveUtils {
+  def withHiveExternalCatalog(sc: SparkContext): SparkContext = {
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:690]: Could you move this method into `TestHiveContext`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala [hunk 5]
```diff
@@ -88,4 +89,17 @@ class HiveSharedStateSuite extends SparkFunSuite {
...
+    val conf = new SparkConf().setMaster("local").setAppName("SPARK-34568")
+    val sc = SparkContext.getOrCreate(conf)
+    val catalog = sc.conf.get(StaticSQLConf.CATALOG_IMPLEMENTATION)
+    sc.conf.set(StaticSQLConf.CATALOG_IMPLEMENTATION, "in-memory")
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala:97]: try ... finally?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:690]: > Could you move this method into `TestHiveContext`?

Sure, more suitable place.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 6]
```diff
@@ -647,6 +647,10 @@ private[hive] object TestHiveContext {
     scratchDir
   }
 
+  def withHiveExternalCatalog(sc: SparkContext): SparkContext = {
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:650]: private

**Comment by maropu** [comment]
Seems fine to me cc: @yaooqinn  @cloud-fan

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala]: > private

Done

**Review by yaooqinn — APPROVED** [review]
looks fine to me, just a few minor revisions

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala [hunk 7]
```diff
@@ -88,4 +89,21 @@ class HiveSharedStateSuite extends SparkFunSuite {
...
+
+  test("SPARK-34568: We should respect enableHiveSupport when initialize SparkSession") {
+    val conf = new SparkConf().setMaster("local").setAppName("SPARK-34568")
+    val sc = SparkContext.getOrCreate(conf)
+    val catalog = sc.conf.get(StaticSQLConf.CATALOG_IMPLEMENTATION)
+    try {
+      sc.conf.set(StaticSQLConf.CATALOG_IMPLEMENTATION, "in-memory")
+      val sparkSession = SparkSession.builder().enableHiveSupport().sparkContext(sc).getOrCreate()
+      assert(
+        sparkSession.sparkContext.conf.get(StaticSQLConf.CATALOG_IMPLEMENTATION) == "in-memory")
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala]: > nit: ===

Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala [hunk 8]
```diff
@@ -88,4 +89,21 @@ class HiveSharedStateSuite extends SparkFunSuite {
     assert(!ss.conf.get(WAREHOUSE_PATH).contains(invalidPath),
       "session level conf should be passed to catalog")
   }
+
+  test("SPARK-34568: We should respect enableHiveSupport when initialize SparkSession") {
+    val conf = new SparkConf().setMaster("local").setAppName("SPARK-34568")
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala:95]: > nit: disable ui

Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala [hunk 9]
```diff
@@ -88,4 +89,21 @@ class HiveSharedStateSuite extends SparkFunSuite {
...
+    val sc = SparkContext.getOrCreate(conf)
+    val catalog = sc.conf.get(StaticSQLConf.CATALOG_IMPLEMENTATION)
+    try {
+      sc.conf.set(StaticSQLConf.CATALOG_IMPLEMENTATION, "in-memory")
+      val sparkSession = SparkSession.builder().enableHiveSupport().sparkContext(sc).getOrCreate()
+      assert(
+        sparkSession.sparkContext.conf.get(StaticSQLConf.CATALOG_IMPLEMENTATION) == "in-memory")
+      assert(sparkSession.sharedState.conf.get(StaticSQLConf.CATALOG_IMPLEMENTATION) == "hive")
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala]: > ditto

Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala [hunk 10]
```diff
@@ -88,4 +89,21 @@ class HiveSharedStateSuite extends SparkFunSuite {
     assert(!ss.conf.get(WAREHOUSE_PATH).contains(invalidPath),
       "session level conf should be passed to catalog")
   }
+
+  test("SPARK-34568: We should respect enableHiveSupport when initialize SparkSession") {
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala:94]: > the test name and the PR title are too general, can we describe the situation?

Done

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**